pkg/controller/clusterinstallation/controller.go

```diff
@@ -118,27 +109,23 @@ func (r *ReconcileClusterInstallation) Reconcile(request reconcile.Request) (rec
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	reqLogger.Info("Reconciling ClusterInstallation")
 
-	// Fetch the ClusterInstallation instance
+	// Fetch the ClusterInstallation.
 	mattermost := &mattermostv1alpha1.ClusterInstallation{}
 	err := r.client.Get(context.TODO(), request.NamespacedName, mattermost)
 	if err != nil && k8sErrors.IsNotFound(err) {
-		// Request object not found, could have been deleted after reconcile request.
-		// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
-		// Return and don't requeue
-		r.setReconciling()
+		// Request object not found, could have been deleted after reconcile
+		// request. Owned objects are automatically garbage collected. For
+		// additional cleanup logic use finalizers. Return and don't requeue.
 		return reconcile.Result{}, nil
 	} else if err != nil {
 		// Error reading the object - requeue the request.
-		r.setReconciling()
 		return reconcile.Result{}, err
 	}
 
-	if mattermost.Status.State != r.state {
-		status := mattermost.Status
-		status.State = r.state
-		err = r.updateStatus(mattermost, status, reqLogger)
+	// Set a new ClusterInstallation's state to reconciling.
+	if len(mattermost.Status.State) == 0 {
+		err = r.setStateReconciling(mattermost, reqLogger)
```

Comment:
Is this the main fix?

Reply:
The main part of the fix is that fact that we no longer keep track of the state as an in-memory var on the reconciler. Instead, we just compare the previous state variable to whatever the reconciler thinks the state should be in that moment. The in-memory state var was how new CIs had the possibility of being set to stable right away. I am surprised it didn't also cause other issues.

This part specifically is just the reconciler setting the state on a new CI that was just created. It should only happen once per custom resource that is created.

Reply:
Cool - got it 👍